DeviceCommunications/DeviceComsProtocolParser.cs

```diff
@@ -0,0 +1,194 @@
+        /// Changing this has no effect if the timer is running.
+        /// </summary>
+        double IEEGData.DigitalMinimum { get; } = -1;
+
+        double IEEGData.DigitalMaximum { get; } = 1;
+
+        double IEEGData.PhysicalMinimum { get; } = -20;
+
+        double IEEGData.PhysicalMaximum { get; } = 20;
+
+        EEGSignalMetadata SignalMetadata { get; } = new EEGSignalMetadata()
+        {
+            Label = "EEG",
+            Units = "mV",
+            TransducerType = "UNKNOWN",
+            Prefiltering = "UNKNOWN",
+        };
+
+        // Returns a copy whenever it's accessed, to avoid concurrency issues.
+        public IEnumerable<(long, EEGData)> DataPoints
+        {
+            get
+            {
+                lock (_dataPointsLock)
+                {
+                    return _dataPoints.ToArray();
+                }
+            }
+        }
+
+        private void DataReceivedEventHandler(object sender, EventArgs e)
+        {
+            byte[] rxData;
+
+            int count = _DeviceComsStream.GetReceivedData(out rxData);
+
+            foreach (byte b in rxData)
+            {
+                _RxData[_RxIndex] = b;
+
+                switch (_RxIndex)
+                {
+                    case 0:
+                        if (0xFF == b)
+                        {
+                            _RxIndex = 1;
+                            break;
+                        }
+
+                        _RxIndex = 0;
+                        break;
+                    case 1:
+                        if (0xFF == b)
+                        {
+                            _RxIndex = 2;
+                            break;
+                        }
+
+                        _RxIndex = 0;
+                        break;
+                    case 2:
+                    case 3:
+                    case 4:
+                    case 5:
+                    case 6:
+                    case 7:
+                    case 8:
+                    case 9:
+                    case 10:
+                    case 11:
+                    case 12:
+                    case 13:
+                    case 14:
+                    case 15:
+                    case 16:
+                    case 17:
+                    case 18:
+                        _RxIndex++;
+                        break;
+                    case 19:
+                        // all data captured so parse it.
+                        long timeMs = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
+
+                        double[] cData = new double[8];
+
+                        for (int i = 0; i < 8; i++)
+                        {
+                            cData[i] = BitConverter.ToInt16(_RxData, ((i * 2) + 4));
+                        }
+
+                        EEGData voltages = new EEGData(cData);
+
+                        lock (_dataPointsLock)
+                        {
+                            _dataPoints.Enqueue((timeMs, voltages));
```

Comment:
Based on our last discussion, I think we'll be wanting to split this apart again and have a separate pipeline per channel, each of which will get a (time, voltage data point) pair - have I got that right?

Reply:
Yeah I think so, I thought we could have a class between this one and the pipelines to distribute the data to each of them.